File: Frontend/SimplifyingPass/src/dev/vale/simplifying/Conversions.scala
package dev.vale.simplifying

import dev.vale.{CodeLocationS, finalast, vimpl}
import dev.vale.finalast._
import dev.vale.postparsing._
import dev.vale.highertyping._
import dev.vale.finalast._
import dev.vale.instantiating.ast._
import dev.vale.postparsing.rules._
import dev.vale.{finalast => m, postparsing => s}

object Conversions {
  def evaluateCodeLocation(loc: CodeLocationS): CodeLocation = {
    val CodeLocationS(line, col) = loc
    finalast.CodeLocation(line, col)
  }

  def evaluateMutability(mutability: MutabilityI): Mutability = {
    mutability match {
      case MutableI => Mutable
      case ImmutableI => Immutable
    }
  }

  def evaluateMutabilityTemplata(mutability: MutabilityI): Mutability = {
    mutability match {
      case MutableI => Mutable
      case ImmutableI => Immutable
    }
  }

  def evaluateVariabilityTemplata(mutability: VariabilityI): Variability = {
    mutability match {
      case VaryingI => Varying
      case FinalI => Final
    }
  }

  def evaluateLocation(location: LocationI): LocationH = {
    location match {
      case InlineI => InlineH
      case YonderI => YonderH
    }
  }

  def evaluateVariability(variability: VariabilityI): Variability = {
    variability match {
      case FinalI => Final
      case VaryingI => Varying
    }
  }

  def evaluateOwnership(ownership: OwnershipI): OwnershipH = {
    ownership match {
      case OwnI => OwnH
      case ImmutableBorrowI => ImmutableBorrowH
      case MutableBorrowI => MutableBorrowH
      case ImmutableShareI => ImmutableShareH
      case MutableShareI => MutableShareH
      case WeakI => WeakH
    }
  }

  def unevaluateTemplataType()(tyype: ITemplataType): ITemplataType = {
    tyype match {
      case CoordTemplataType() => CoordTemplataType()
      case KindTemplataType() => KindTemplataType()
      case IntegerTemplataType() => IntegerTemplataType()
      case BooleanTemplataType() => BooleanTemplataType()
      case MutabilityTemplataType() => MutabilityTemplataType()
      case LocationTemplataType() => LocationTemplataType()
      case OwnershipTemplataType() => OwnershipTemplataType()
      case VariabilityTemplataType() => VariabilityTemplataType()
      case TemplateTemplataType(_, _) => vimpl() // can we even specify template types in the syntax?
    }
  }
}

File: FrontendRust/src/simplifying/conversions.rs
#[allow(unused_imports)]
use crate::utils::range::CodeLocationS;
#[allow(unused_imports)]
use crate::instantiating::ast::types::{MutabilityI, VariabilityI, OwnershipI, LocationI};
#[allow(unused_imports)]
use crate::final_ast::types::{CodeLocation, Mutability, Variability, OwnershipH, LocationH};
#[allow(unused_imports)]
use crate::postparsing::itemplatatype::ITemplataType;

pub fn evaluate_code_location(loc: CodeLocationS) -> CodeLocation {
    panic!("Unimplemented: evaluate_code_location");
}

pub fn evaluate_mutability(mutability: MutabilityI) -> Mutability {
    match mutability {
        MutabilityI::Mutable => Mutability::Mutable,
        MutabilityI::Immutable => Mutability::Immutable,
    }
}

pub fn evaluate_mutability_templata(mutability: MutabilityI) -> Mutability {
    match mutability {
        MutabilityI::Mutable => Mutability::Mutable,
        MutabilityI::Immutable => Mutability::Immutable,
    }
}

pub fn evaluate_variability_templata(mutability: VariabilityI) -> Variability {
    match mutability {
        VariabilityI::Varying => Variability::Varying,
        VariabilityI::Final => Variability::Final,
    }
}

pub fn evaluate_location(location: LocationI) -> LocationH {
    panic!("Unimplemented: evaluate_location");
}

pub fn evaluate_variability(variability: VariabilityI) -> Variability {
    match variability {
        VariabilityI::Final => Variability::Final,
        VariabilityI::Varying => Variability::Varying,
    }
}

pub fn evaluate_ownership(ownership: OwnershipI) -> OwnershipH {
    match ownership {
        OwnershipI::Own => OwnershipH::OwnH,
        OwnershipI::ImmutableBorrow => OwnershipH::ImmutableBorrowH,
        OwnershipI::MutableBorrow => OwnershipH::MutableBorrowH,
        OwnershipI::ImmutableShare => OwnershipH::ImmutableShareH,
        OwnershipI::MutableShare => OwnershipH::MutableShareH,
        OwnershipI::Weak => OwnershipH::WeakH,
    }
}

pub fn unevaluate_templata_type(tyype: ITemplataType) -> ITemplataType {
    panic!("Unimplemented: unevaluate_templata_type");
}

